src/addin/addin-client.ts

```diff
@@ -316,11 +328,11 @@ export class AddinClient {
   }
 
   /**
-   * Regiesters a callback to be executed when the specified event type occurs.
+   * Registers a callback to be executed when the specified event type occurs.
    * @param eventType The event type to process.
-   * @param callback The callback to executes when the event occurs.
+   * @param callback The callback to execute when the event occurs.
    */
-  public addEventHandler(eventType: string, callback: (context: any) => void) {
+  public addEventHandler(eventType: string, callback: AddinEventCallback) {
```

Comment:
I know we're already using callbacks for the gift batch work, so this is probably a perfectly working solution, but I'm wondering if callbacks are now considered out-of-fashion in favor of more async approaches (like Promises). I'm legitimately asking; I know Promises (and now Observables) have replaced callbacks for most modern JS/TS APIs, but maybe callbacks are still considered OK for this type of implementation (where the caller will eventually call `done()`).

The thought just occurred to me that we may want to resolve the event (`this.postEventReceivedMessage(message.eventRequestId)`) when something like a Promise completed vs relying on the add-in to 'call' us back (done()). But maybe there is weirdness either way.

Separate question - do you think there is any value to having first-class methods for adding each handler type? I can see minor benefits to it--like avoiding event type typos, but I was wondering if we may need event-type-specific handler signatures in the future (for example, if we have a `pre-save` event type where the callback/promise needs to return a `ValidationResult` (making that up)). I still think we may have random one-off event types that individual forms support where we'd need a generic `addEventHandler`, like this. But having convenience methods for common event types may be nice.

Reply:
Yeah, I thought about using a more modern approach to handling events.  Promise seems to be used for a single async task - kick it off and wait to finish or error, instead of what I need here which is async with n multiple values to be returned.  Rxjs observable would do the trick, but not really sure we want to pull this in for our simple vanilla JS client lib.

> The thought just occurred to me that we may want to resolve the event when something like a Promise completed vs relying on the add-in to 'call' us back (done())

Correct me if I'm missing something, but I think this would only be the case only if we returned a Promise from the `addEventHandler` which I'm not thinking is the right approach over a callback.

As far as the first class methods, I was mostly avoiding doing that because I didn't want to have too many form-specific methods added to the AddinClient.  I think a "Save" event is very specific to forms - but maybe there's others I'm not thinking of.  Whereas the generic `addEventHandler` could be used for other types of add-ins.
But maybe we could have a "FormEvents" property on the AddinClient that returns a class containing the additional form-specific event handlers.

Reply:
> Correct me if I'm missing something, but I think this would only be the case only if we returned a Promise from the addEventHandler which I'm not thinking is the right approach over a callback.

Yeah, I was thinking about some async task that gets executed each time the event is fired, and `processHostEvent` would wait for that to complete. You're right, a single promise configured as the event handler wouldn't work... maybe if the callback itself returned a promise that we waited to resolve.   

I'm fine with this approach though. Was just wondering about the "modern" approach (which I admittedly don't know). I kind of like the `FormEvents` idea (maybe a property of `formAddinConfig`--in case we ever have other form-addin-specific properties?), but having the same events API regardless is type is good too. I'll stop slowing progress, proceed! :)

Reply:
Ah, I see what you mean about the Promise now.  So instead of `export type AddinEventCallback = (context: any, done?: () => void) => void;` containing a `done` callback to be called, instead that second param could be a promise to resolve.  I like that, and it seems more modern.  Let me try that out to see if it'll work.

Reply:
@Blackbaud-ChristopherRodgers I looked into a couple places where a Promise could be used, and I pushed up some changes to resolve a promise, when add-in client processing is done, that will respond to the host that the client event has been received/processed